### PR TITLE
add focal loss

### LINKED ISSUE #39: add focal loss and unit test
**Describe the solution you'd like**
ATT.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by GaryGao99**
Fix #39 

**Thank you for submitting a pull request! Please provide the following information for code review:**

# Pull Request Summary
[TODO] add focal loss . #39 

# Test Plan
[TODO] python loss_utils_test.py

# Reviewers
[TODO] @zh794390558

### CODE REVIEW

File: delta/utils/loss/loss_impl.py [hunk 1]
```diff
@@ -166,3 +166,24 @@ def call(self,
...
+           input_length=None,
+           labels=None,
+           label_length=None,
+           soft_labels=None):
+
+    del input_length
+    del label_length
+    del soft_labels
+
+    return focal_loss(
```
Comment (hankun11) [inline delta/utils/loss/loss_impl.py]: `from delta.utils.loss.loss_utils import focal_loss` is missing?

File: delta/utils/loss/loss_utils_test.py [hunk 2]
```diff
@@ -183,6 +183,14 @@ def gen_fake_data(batch_size, embedding_size, num_spks):
                                        limit_to_pi=True)
       self.assertAllClose(output.eval(), output_true)
 
+  def test_focal_loss(self):
+    print('test_focal_loss')
+    with self.cached_session():
+      loss = loss_utils.focal_loss(
+          logits=tf.constant(self.logits),
+          labels=tf.constant(self.labels))
+      self.assertAllClose(loss.eval(), 0.0, rtol=1e-06, atol=1e-6)
```
Comment (hankun11) [inline delta/utils/loss/loss_utils_test.py]: For unit test, I think that it is better to add a test to show the difference between the focal loss and CE, e.g., easy examples get lower losses. Specifically, you may compute the **non-zero loss** to show the correctness, similar to #41  
Thanks!

File: delta/utils/loss/loss_impl.py [hunk 3]
```diff
@@ -166,3 +167,24 @@ def call(self,
...
+class FocalLoss(Loss):
+  def __init__(self, config):
+    super().__init__(config)
+
+  def call(self,
+           logits=None,
+           input_length=None,
+           labels=None,
+           label_length=None,
+           soft_labels=None):
```
Comment (zh794390558) [inline delta/utils/loss/loss_impl.py]: **kwargs)

File: delta/utils/loss/loss_utils.py [hunk 4]
```diff
@@ -200,3 +200,28 @@ def arcface_loss(embedding,
         tf.multiply(cos_mt_temp, mask),
         name='arcface_loss_output')
   return output
+
+def focal_loss(logits, labels, gamma=2, name='focal_loss'):
+    """
+    Focal loss for binary classification
+    :param logits: A float32 tensor of shape [batch_size].
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils.py]: shape [batch_size, num_class]

File: delta/utils/loss/loss_utils_test.py [hunk 5]
```diff
@@ -183,6 +183,14 @@ def gen_fake_data(batch_size, embedding_size, num_spks):
                                        limit_to_pi=True)
       self.assertAllClose(output.eval(), output_true)
 
+  def test_focal_loss(self):
+    print('test_focal_loss')
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils_test.py]: logging.info

File: delta/utils/loss/loss_utils.py [hunk 6]
```diff
@@ -200,3 +200,28 @@ def arcface_loss(embedding,
         tf.multiply(cos_mt_temp, mask),
         name='arcface_loss_output')
   return output
+
+def focal_loss(logits, labels, gamma=2, name='focal_loss'):
+    """
+    Focal loss for binary classification
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils.py]: binary classification?

File: delta/utils/loss/loss_utils_test.py [hunk 7]
```diff
@@ -183,6 +183,14 @@ def gen_fake_data(batch_size, embedding_size, num_spks):
...
+    with self.cached_session():
+      loss = loss_utils.focal_loss(
+          logits=tf.constant(self.logits),
+          labels=tf.constant(self.labels))
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils_test.py]: using specific gamma param

File: delta/utils/loss/loss_utils.py [hunk 8]
```diff
@@ -200,3 +200,28 @@ def arcface_loss(embedding,
...
+    :param logits: A float32 tensor of shape [batch_size].
+    :param labels: A int32 tensor of shape [batch_size].
+    :param gamma: A scalar for focal loss gamma hyper-parameter.
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils.py]: gamma must large than zero

File: delta/utils/loss/loss_utils.py [hunk 9]
```diff
@@ -200,3 +200,28 @@ def arcface_loss(embedding,
         tf.multiply(cos_mt_temp, mask),
         name='arcface_loss_output')
   return output
+
+def focal_loss(logits, labels, gamma=2, name='focal_loss'):
+    """
+    Focal loss for multi classification
+    :param logits: A float32 tensor of shape [batch_size num_class].
+    :param labels: A int32 tensor of shape [batch_size, num_class].
```
Comment (zh794390558) [inline delta/utils/loss/loss_utils.py]: [batch_size, num_class] or [batch_size]

File: delta/utils/loss/loss_impl.py [hunk 10]
```diff
@@ -161,8 +159,33 @@ def call(self,
...
     loss = mask_sequence_loss(logits, labels, input_length, label_length)
     return loss
+
+@registers.loss.register
+class FocalLoss(Loss):
+  def __init__(self, config):
+    super().__init__(config)
+
+    self.gamma = 2
+    if 'gamma' in self._config['solver']['optimizer']:
```
Comment (zh794390558) [inline delta/utils/loss/loss_impl.py]: ['solver']['optimizer']['loss']
